### PR TITLE
Handle root, network path or disk drive letter

### PR DESCRIPTION
**Pull request by hun1ahpu**
Fix for #19382 and #19384

Always preserve root path (/root/path), network path (\\network\path) or disk drive (diskDrive:\path)

Windows tests are passing. Could anyone please check if non windows tests are also passing?

### CODE REVIEW

**Comment by bpasero** [comment]
They seem not to: 

```
1) Labels shorten - not windows:

      AssertionError: [ '.', 'a', 'b', 'b/c', 'a/c' ] deepEqual [ '', 'a', 'b', 'b/c', 'a/c' ]
      + expected - actual

       [
      -  "."
      +  ""
         "a"
         "b"
         "b/c"
         "a/c"
      
      at Context.<anonymous> (out/vs/base/test/common/labels.test.js:83:20)
```

**Comment by hun1ahpu** [comment]
My bad. Forgot to update 1 test.

File: src/vs/base/common/labels.ts [hunk 1]
```diff
@@ -88,42 +88,82 @@ export function shorten(paths: string[]): string[] {
...
+		if (!path) {
+			shortenedPaths[pathIndex] = path;
+			continue;
+		}
+
 		match = true;
 
+		let prefix = '';
+		// trim for now and concatenate network path (e.g. \\network) or root path (/etc) later
+		if (path.indexOf(`${nativeSep + nativeSep}`) === 0) {
```
Comment (bpasero) [inline src/vs/base/common/labels.ts]: I suggest to extract a new `const unc = ${nativeSep + nativeSep}`

File: src/vs/base/common/labels.ts [hunk 2]
```diff
@@ -88,42 +88,82 @@ export function shorten(paths: string[]): string[] {
...
+			prefix = path.substr(0, path.indexOf(`${nativeSep + nativeSep}`) + 2);
+			path = path.substr(path.indexOf(`${nativeSep + nativeSep}`) + 2);
+		} else if (path.indexOf(`${nativeSep}`) === 0) {
+			prefix = path.substr(0, path.indexOf(`${nativeSep}`) + 1);
```
Comment (bpasero) [inline src/vs/base/common/labels.ts]: Makes it unnecessarily hard to read to put a variable around ticks, just use `nativeSep` here?

**Review by bpasero — COMMENTED** [review]
@hun1ahpu I tested this on macOS and Windows and I think I am not seeing the drive letter, nor root slash in the label when I have duplicates:

image

image

Isn't that the idea?

Also, can't this change be implemented much simpler? My understanding is that we want the current behaviour but we just want to append either "." (for empty label) or the drive letter/slash/unc in the beginning. Why all the code changes needed if we just need to append the stuff after we computed the label?

E.g.

`.../something/...` becomes `/.../something/...` (on Windows using drive letter and backslash).

**Comment by hun1ahpu** [comment]
@bpasero I thought the idea was to display the root/disk drive only for the files outside of the opened folder. All the rest should keep the same behavior and display only the relative paths.
@Tyriar to confirm if my understanding is correct.

Here is a screenshot for example:
image

The logic becomes more complicated due to the several corner cases:
1) c:\a\b, d:\a\b. I'd expect to have c:\... and d:\... as a result
2) c:\a\d, c:\a\b. I'd expect to have c:\...\d, c:\...\b as a result
3) c:\a\b, c:\d\e. I'd expect to have c:\a\..., c:\d\... as a result

Let me know if that's not what you'd expect.

Once we agree on the expected behavior i will take a look and try to simplify the logic if I can.

Thank you for your feedback.

**Comment by Tyriar** [comment]
@hun1ahpu yes that was the idea, @bpasero may have a different folder open though (I see "monaco" in the title bar, not "vscode"). If they are out of folder I would expect "/.../terminal/..." and "C:\...\terminal\..." in his screenshot.

**Comment by bpasero** [comment]
Ah ok, yes all those files where from the same workspace and not outside. So for those the behaviour should be the same, we only change it for "out of workspace" files?

**Comment by Tyriar** [comment]
Yes only out of workspace so it's obvious say if you open ./gulpfile.js and ../some-other-folder/gulpfile.js

**Comment by hun1ahpu** [comment]
I will go forward and address the comments from @bpasero and will try to simplify the code if possible if the overall behavior works for you guys?

**Comment by Tyriar** [comment]
The tests look good :+1:

**Comment by bpasero** [comment]
Pretty cool, thanks!

**Pull request merged**